crates/sui-network/src/state_sync/mod.rs

```diff
@@ -1046,6 +1068,75 @@ where
     Ok(())
 }
 
+async fn sync_checkpoint_contents_from_archive<S>(
+    network: anemo::Network,
+    archive_readers: ArchiveReaderBalancer,
+    store: S,
+    peer_heights: Arc<RwLock<PeerHeights>>,
+    metrics: Metrics,
+) where
+    S: WriteStore + Clone + Send + Sync + 'static,
+    <S as ReadStore>::Error: std::error::Error,
+{
+    loop {
+        let peers: Vec<_> = peer_heights
+            .read()
+            .unwrap()
+            .peers_on_same_chain()
+            // Filter out any peers who we aren't connected with.
+            .filter_map(|(peer_id, info)| network.peer(*peer_id).map(|peer| (peer, *info)))
+            .collect();
+        let lowest_checkpoint_on_peers = peers
+            .iter()
+            .map(|(_p, state_sync_info)| state_sync_info.lowest)
+            .min();
+        let highest_synced = store
+            .get_highest_synced_checkpoint()
+            .expect("store operation should not fail")
+            .sequence_number;
+        let sync_from_archive = if let Some(lowest_checkpoint_on_peers) = lowest_checkpoint_on_peers
+        {
+            highest_synced < lowest_checkpoint_on_peers
+        } else {
+            false
+        };
+        if sync_from_archive {
+            let start = highest_synced
+                .checked_add(1)
+                .expect("Checkpoint seq num overflow");
+            if let Some(archive_reader) = archive_readers
+                .pick_one_random(start..lowest_checkpoint_on_peers.unwrap())
+                .await
+            {
+                let txn_counter = Arc::new(AtomicU64::new(0));
+                let checkpoint_counter = Arc::new(AtomicU64::new(0));
+                let (_sender, receiver) = oneshot::channel();
+                tokio::spawn(update_highest_sync_metrics(
+                    receiver,
+                    store.clone(),
+                    metrics.clone(),
+                ));
```

Comment:
i think it would be cleaner if the reader updated the metrics itself rather than doing this async. Not a huge deal though

Reply:
for sure, i will try to integrate it in the reader itself

Reply:
This turns out leads to cyclic dependencies b/w `sui-archival` and `sui-network` crates (currently sui-network has a dependency on sui-archival). It could be refactored, but leaving is as is for now since it is not a critical requirement.